### PR TITLE
Small allocation optimization in StateTable.Builder.ToImmutable

### PR DESCRIPTION
**Pull request by ToddGrun**
Use an ArrayBuilder as a scratch buffer instead of allocating an array every time. Small allocation that barely showed up in the RichCopyTest speedometer test. 

*** Previous allocations
image

### CODE REVIEW

**Comment by ToddGrun** [comment]
Debated not making this change, but it was simple and safe enough that I figured it was worth it

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/StateTableStore.cs [hunk 1]
```diff
@@ -51,10 +52,15 @@ public sealed class Builder
             public StateTableStore ToImmutable()
             {
                 // we can cache the tables at this point, as we'll no longer be using them to determine current state
-                var keys = _tableBuilder.Keys.ToArray();
-                foreach (var key in keys)
+                var scratch = ArrayBuilder<KeyValuePair<object, IStateTable>>.GetInstance(_tableBuilder.Keys.Count);
+                foreach (var kvp in _tableBuilder)
                 {
-                    _tableBuilder[key] = _tableBuilder[key].AsCached();
+                    scratch.Add(kvp);
```
Comment (sharwell) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/StateTableStore.cs:58]: 💡 No need to add if already cached

```suggestion
                    var cached = kvp.Value.AsCached();
                    if (cached != kvp.Value)
                        scratch.Add(new KeyValuePair<object, IStateTable>(kvp.Key, cached));
```

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/StateTableStore.cs [hunk 2]
```diff
@@ -51,10 +52,15 @@ public sealed class Builder
             public StateTableStore ToImmutable()
             {
                 // we can cache the tables at this point, as we'll no longer be using them to determine current state
-                var keys = _tableBuilder.Keys.ToArray();
```
[the remark this answers is not in the document]
Comment (ToddGrun) [reply src/Compilers/Core/Portable/SourceGeneration/Nodes/StateTableStore.cs:54]: As mentioned offline, the scratch array would be unnecessary if there were a CollectionsMarshal.GetValueRefOrNullRef implementation for ImmutableSegmentedDictionary. As there is not, I'd like to move forward with these changes (I did notice that when there are no entries in _tableBuilder, that the enumerator still creates the mutable dictionary, but that's a separate issue in the ImmutableSegmentedDictionary implementation).

**Comment by ToddGrun** [comment]
@dotnet/roslyn-compiler -- ptal

**Comment by ToddGrun** [comment]
> waiting on #72395 as it will provide a cleaner way to do this

Modified this PR to now use SegmentedCollectionsMarshal.GetValueRefOrNullRef instead of a scratch array.

**Review by sharwell — APPROVED** [review]

**Review by 333fred — APPROVED** [review]

**Review by jjonescz — APPROVED** [review]

**Pull request merged**